src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualCharSequence.Chunks.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars
+{
+    internal partial struct VirtualCharSequence
+    {
+        /// <summary>
+        /// Abstraction over a contiguous chunk of <see cref="VirtualChar"/>s.  This
+        /// is used so we can expose <see cref="VirtualChar"/>s over an <see cref="ImmutableArray{VirtualChar}"/>
+        /// or over a <see cref="string"/>.  The latter is especially useful for reducing
+        /// memory usage in common cases of string tokens without escapes.
+        /// </summary>
+        private abstract partial class Chunk
+        {
+            protected Chunk()
+            {
+            }
+
+            public abstract int Length { get; }
+            public abstract VirtualChar this[int index] { get; }
+        }
+
+        /// <summary>
+        /// Thin wrapper over an actual <see cref="ImmutableArray{VirtualChar}"/>.
+        /// This will be the common construct we generate when getting the
+        /// <see cref="Chunk"/> for a string token that has escapes in it.
+        /// </summary>
+        private class ImmutableArrayChunk : Chunk
+        {
+            private readonly ImmutableArray<VirtualChar> _array;
+
+            public ImmutableArrayChunk(ImmutableArray<VirtualChar> array)
+                => _array = array;
+
+            public override int Length => _array.Length;
+            public override VirtualChar this[int index] => _array[index];
+        }
+
+        /// <summary>
+        /// Represents a <see cref="Chunk"/> on top of a normal
+        /// string.  This is the common case of the type of the sequence we would
+        /// create for a normal string token without any escapes in it.
+        /// </summary>
+        private class StringChunk : Chunk
+        {
+            private readonly int _firstVirtualCharPosition;
+
+            /// <summary>
+            /// The underlying string that we're returning virtual chars from.
+            /// Note the chars we return will normally be from a subsection of this string.
+            /// i.e. the _underlyingData will be something like:  "abc" (including the quotes).
+            /// The <see cref="_underlyingDataSpan"/> will snip out the quotes, leaving just
+            /// "abc"
+            /// </summary>
+            private readonly string _underlyingData;
```

Comment:
>_underlyingData [](start = 36, length = 15)

`_underlyingData` and `_underlyingDataSpan` taken together seem to form a `Span<char>`. Is that an option?

Reply:
@jcouv i will see!

Reply:
@jcouv It says i would need a ref-struct for that.  however, i have to use a class here, because hte point is to subclass Chunk and provide alternative implementations.

Reply:
Ah, good point. Thanks

---
In reply to: [262249499](https://github.com/dotnet/roslyn/pull/33834#discussion_r262249499) [](ancestors = 262249499)

Reply:
Isn't that what `ReadOnlyMemory<char>` is for? When you want a `Span<T>` but need something that's not a `ref struct`.

Reply:
Note: this changed anyways, since we no longer keep a span here.  So this is somewhat moot.

However, i do believe that ReadOnlyMemory would be appropriate here.  However, investigating more i could not find the ncessary extensions to allow us to use this.  Perhaps they're in a version of the bcl libraries we're not referencing yet?  Specifically, i cannot find `SpanExtensions.AsReadOnlyMemory(this string s)`.